metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcher.scala

```diff
@@ -117,89 +117,140 @@ object FileWatcher {
-        scribe.debug(s"Registering root for file watching: $root")
-        repo.register(root, Int.MaxValue)
+  )(implicit ec: ExecutionContext): () => Unit = {
+    val watchEventQueue: BlockingQueue[FileWatcherEvent] =
+      new LinkedBlockingQueue[FileWatcherEvent]
+
+    val watcher: PathWatcher[PathWatchers.Event] = {
+      if (scala.util.Properties.isMac) {
+        // Due to a hard limit on the number of FSEvents streams that can be
+        // opened on macOS, only up to 32 longest common prefixes of the files to
+        // watch are registered for a recursive watch.
+        // However, the events are then filtered to receive only relevant events
+
+        val trie = PathTrie(
+          pathsToWatch.files ++ pathsToWatch.directories
+        )
+        val isWatched = trie.containsPrefixOf _
+
+        // Select up to `maxRoots` longest prefixes of all files in the trie for
+        // watching. Watching the root of the workspace may have bad performance
+        // implications if it contains many other projects that we don't need to
+        // watch (eg. in a monorepo)
+        val watchRoots =
+          trie.longestPrefixes(workspace.getRoot(), 1)
+
+        val watcher = initWatcher(
+          path => watchFilter(path) && isWatched(path),
+          watchEventQueue
+        )
+        watchRoots.foreach { root =>
+          scribe.debug(s"Registering root for file watching: $root")
+          watcher.register(root, Int.MaxValue)
+        }
+
+        watcher
+      } else {
+        // // Other OSes register all the files and directories individually
+        val watcher = initWatcher(watchFilter, watchEventQueue)
+
+        pathsToWatch.directories.foreach(
+          watcher.register(_, Int.MaxValue)
+        )
+        pathsToWatch.files.foreach(watcher.register(_, -1))
+
+        watcher
       }
-      () => repo.close()
-    } else {
-      // Other OSes register all the files and directories individually
-      val repo = initFileTreeRepository(watchFilter, callback)
+    }
 
-      filesToWatch.sourceDirectories.foreach(repo.register(_, Int.MaxValue))
-      filesToWatch.semanticdDirectories.foreach(repo.register(_, Int.MaxValue))
-      filesToWatch.sourceFiles.foreach(repo.register(_, -1))
+    val stopWatchingSignal: AtomicBoolean = new AtomicBoolean
 
-      () => repo.close()
+    // queue is processed serially to defensively prevent possible races
+    // when a file is created and then deleted in quick succesion.
+    // Upstream callbacks implicitly assume that events for a single file
+    // arrive and are processed serially.
+    // This consumption could be parallelized, but in case in indexing,
+    // compilation is usually the dominating bottleneck
+    repeatUntilStopped(
+      stopWatchingSignal, {
+        Future {
+          blocking {
+            // do not block indefinitely to allow the loop to be stopped by stop signal
+            Option(watchEventQueue.poll(200, TimeUnit.MILLISECONDS))
```

Comment:
It isn't maybe idiomatic scala, but the way I like to do something like this is:
```scala
val thread = new Thread("metals-watch-callback-thread") {
  @tailrec def loop(): Unit = {
    try callback(watchEventQueue.take)
    catch { case _: InterruptedException =>  }
    if (!stopWatchingSignal.get) loop()
  }
  override def run(): Unit = loop()
}
...
() => {
  watcher.close()
  stopWatchingSignal.set(true)
  thread.interrupt()
}
```

It's a minor thing but with this technique the watch thread stays idle until there really is work to do and it will also shutdown essentially instantaneously instead of on average taking 100ms.

Reply:
I prefer this version :) `watchEventQueue.take` will block until there's an event in the queue and the watch thread stays idle 😍

Reply:
changed in a fixup commit